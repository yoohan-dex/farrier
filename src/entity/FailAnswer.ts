import { Entity, PrimaryGeneratedColumn, Column } from 'typeorm';

@Entity()
export class FailAnswer {
  @PrimaryGeneratedColumn() id: number;

  @Column() questionId: number;

  @Column('simple-array') selectedAnswer: number[];
}
