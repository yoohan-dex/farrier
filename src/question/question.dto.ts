export class CreateSubQuestionDto {
  type: 0;
  content: string;
  answers: string[];
  rightAnswer: string[];
  analysis: string;
}

export class CreateCommonQuestionDto {
  type: 1;
  content: string;
  answers: string[];
  rightAnswer: string[];
  analysis: string;
}

export class CreateSecondQuestionDto {
  type: 2;
  content: string;
  questions: CreateSubQuestionDto[];
}

export class CreateThirdQuestionDto {
  type: 3;
  answers: string[];
  questions: {
    type: number;
    content: string;
    rightAnswer: string[];
    analysis: string;
  }[];
}

export class EstimateCommonQuestionDto {
  type: 1;
  questionId: number;
  selectedAnswers: number[];
}

export class EstimateSecondQuestionDto {
  type: 2;
  questionId: number;
  subQuestions: {
    subQuestionId: number;
    selectedAnswers: number[];
  }[];
}

export class EstimateThirdQuestionDto {
  type: 3;
  questionId: number;
  subQuestions: {
    subQuestionId: number;
    selectedAnswers: number[];
  }[];
}
