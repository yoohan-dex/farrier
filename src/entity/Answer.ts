import { Entity, PrimaryGeneratedColumn, Column, ManyToOne } from 'typeorm';
import { Question } from './Question';

@Entity()
export class Answer {
  @PrimaryGeneratedColumn() id: number;

  @Column() content: string;

  @ManyToOne(type => Question)
  question: Question;
}
