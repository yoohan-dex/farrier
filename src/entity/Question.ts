import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  OneToMany,
  ManyToOne,
} from 'typeorm';
import { Answer } from './Answer';

@Entity()
export class Question {
  @PrimaryGeneratedColumn() id: number;

  @Column('text', { nullable: true })
  content: string;

  @Column('int', { nullable: true })
  type: number;

  @OneToMany(type => Answer, answer => answer.question, {
    eager: true,
    nullable: true,
  })
  answers: Answer[];

  @Column('simple-array', { nullable: true })
  rightAnswer: number[];

  @ManyToOne(type => Question, q => q.questions, { nullable: true })
  parent: Question;

  @OneToMany(type => Question, q => q.parent, { nullable: true })
  questions: Question[];

  @Column('text', { nullable: true })
  analysis: string;
}
