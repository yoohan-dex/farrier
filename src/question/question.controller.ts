import { Controller, Get, Param, Query } from '@nestjs/common';
import { QuestionService } from './question.service';
import { parseFiles } from '../questionsInput/parse';
import { User } from '../shared/decorators/user';
import { WxUser } from '../entity/WxUser';
import { PerformanceService } from '../performance/performance.service';

@Controller('question')
export class QuestionController {
  constructor(
    private readonly questionService: QuestionService,
    private readonly performanceService: PerformanceService,
  ) {}

  @Get()
  async getQuestionByType(
    @Query('type') type: number,
    @Query('page') page: number,
    @Query('size') size: number,
  ) {
    return await this.questionService.getQuestionByType(type, page, size);
  }

  @Get('normal')
  async getQuestionNormal(@User() user: WxUser) {
    const haveDoneIds = await this.performanceService.findDone(user.uuid);
    return await this.questionService.getQuestionNormal(haveDoneIds);
  }

  @Get('trigger/create')
  async triggerCreateQuestions() {
    const questions = await this.questionService.getFileNParse();
    return await this.questionService.saveAllQuestions(questions);
  }
}
