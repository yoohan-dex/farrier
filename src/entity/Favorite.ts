import { Entity, PrimaryGeneratedColumn, ManyToMany, Column } from 'typeorm';
import { Question } from './Question';

@Entity()
export class Favorite {
  @PrimaryGeneratedColumn() id: number;

  @ManyToMany(type => Question, { eager: true })
  questions: Question[];

  @Column() userId: string;
}
