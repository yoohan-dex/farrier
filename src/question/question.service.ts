import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, FindManyOptions, Not, In } from 'typeorm';
import * as fs from 'fs';
import * as path from 'path';

import { Question } from '../entity/Question';
import {
  CreateCommonQuestionDto,
  CreateSecondQuestionDto,
  CreateThirdQuestionDto,
  EstimateCommonQuestionDto,
  EstimateSecondQuestionDto,
  EstimateThirdQuestionDto,
  CreateSubQuestionDto,
} from './question.dto';
import { Answer } from '../entity/Answer';
import { parseFiles } from '../questionsInput/parse';

@Injectable()
export class QuestionService {
  constructor(
    @InjectRepository(Question)
    private readonly questionRepository: Repository<Question>,
    @InjectRepository(Answer)
    private readonly anwserRepository: Repository<Answer>,
  ) {}

  async getQuestionByType(
    type: number,
    page: number = 1,
    size: number = 20,
  ): Promise<Question[]> {
    const options: FindManyOptions<Question> = {
      take: size,
      skip: page ? (page - 1) * 10 : 0,
      relations: ['questions'],
      order: {
        id: 'DESC',
      },
      where: { type },
    };
    const questions = await this.questionRepository.find(options);
    return questions.map(question => ({
      ...question,
      questions:
        question.questions &&
        question.questions.map(subQuestion => ({
          ...subQuestion,
          rightAnswer: subQuestion.rightAnswer.map(
            rightAnswer => 1 * rightAnswer,
          ),
        })),
    }));
  }

  async getQuestionNormal(haveDone: number[] = []) {
    const where = haveDone.length > 0 ? { id: Not(In(haveDone)) } : {};
    const type1query = this.questionRepository.find({
      where: { ...where, type: 1 },
      take: 10,
    });
    const type2query = this.questionRepository.find({
      where: { ...where, type: 2 },
      relations: ['questions'],
      take: 5,
    });
    const type3query = this.questionRepository.find({
      where: { ...where, type: 3 },
      relations: ['questions'],
      take: 5,
    });
    const [type1, type2, type3] = await Promise.all([
      type1query,
      type2query,
      type3query,
    ]);
    return [...type1, ...type2, ...type3];
  }

  async saveCommonQuestion(
    createCommonQuestion:
      | CreateCommonQuestionDto & { parent?: Question }
      | CreateSubQuestionDto,
  ) {
    const answers = await this.saveAnswer(createCommonQuestion.answers);
    const rightAnswer = answers
      .filter(a => createCommonQuestion.rightAnswer.some(v => a.content === v))
      .map(a => a.id);
    const question = new Question();

    question.type = createCommonQuestion.type;
    question.content = createCommonQuestion.content;
    question.answers = answers;
    question.rightAnswer = rightAnswer;
    question.analysis = createCommonQuestion.analysis;
    if (createCommonQuestion.type === 1) {
      question.parent = createCommonQuestion.parent || null;
    }

    return await this.questionRepository.save(question);
  }

  async saveSecondQuestion(createSecondQuestion: CreateSecondQuestionDto) {
    const willSaveQuestion = new Question();
    willSaveQuestion.type = createSecondQuestion.type;
    willSaveQuestion.content = createSecondQuestion.content;
    // const question = await this.questionRepository.save(willSaveQuestion);

    const willSaveSubQuestions = createSecondQuestion.questions.map(async q => {
      const answers = await this.saveAnswer(q.answers);
      const rightAnswer = answers
        .filter(a => q.rightAnswer.some(v => a.content === v))
        .map(a => a.id);
      const willSaveSubQuestion = new Question();
      willSaveSubQuestion.type = 0;
      willSaveSubQuestion.content = q.content;
      willSaveSubQuestion.answers = answers;
      willSaveSubQuestion.rightAnswer = rightAnswer;
      willSaveSubQuestion.analysis = q.analysis;
      return this.questionRepository.save(willSaveSubQuestion);
    });
    const subQuestions = [];
    for (const item of willSaveSubQuestions) {
      subQuestions.push(await item);
    }
    willSaveQuestion.questions = subQuestions;
    const question = await this.questionRepository.save(willSaveQuestion);
    return question;
  }

  async saveThirdQuestion(createThirdQuestion: CreateThirdQuestionDto) {
    const answers = await this.saveAnswer(createThirdQuestion.answers);
    const willSaveSubQuestions = createThirdQuestion.questions.map(q => {
      const question = new Question();
      question.type = q.type;
      question.content = q.content;
      question.analysis = q.analysis;
      question.rightAnswer = answers
        .filter(a => q.rightAnswer.some(v => a.content === v))
        .map(a => a.id);
      return question;
    });
    const subQuestions = await this.questionRepository.save(
      willSaveSubQuestions,
    );

    const question = new Question();
    question.type = createThirdQuestion.type;
    question.answers = answers;
    question.questions = subQuestions;

    return await this.questionRepository.save(question);
  }

  async estimateCommonQuestion(dto: EstimateCommonQuestionDto) {
    const { selectedAnswers } = dto;
    console.log('dto', dto);
    const question = await this.questionRepository.findOne(dto.questionId);
    let allTrue = question.rightAnswer.length === selectedAnswers.length;

    for (const selectedAnswer of selectedAnswers) {
      if (
        !question.rightAnswer
          .map(rightAnswer => rightAnswer * 1)
          .includes(selectedAnswer)
      ) {
        allTrue = false;
      }
    }
    question.rightAnswer = question.rightAnswer.map(v => v * 1);
    return {
      pass: allTrue,
      question,
      selectedAnswer: [
        {
          questionId: question.id,
          selectedAnswers,
        },
      ],
    };
  }

  async estimateSecondQuestion(
    dto: EstimateSecondQuestionDto | EstimateThirdQuestionDto,
  ) {
    const subQuestionIds = dto.subQuestions.map(
      subQuestionDto => subQuestionDto.subQuestionId,
    );
    const subQuestions = await this.questionRepository.findByIds(
      subQuestionIds,
    );
    const pass = subQuestions
      .map((subQuestion, idx) => {
        const { selectedAnswers } = dto.subQuestions[idx];
        let allTrue = subQuestion.rightAnswer.length === selectedAnswers.length;

        for (const selectedAnswer of selectedAnswers) {
          if (
            !subQuestion.rightAnswer
              .map(rightAnswer => rightAnswer)
              .includes(selectedAnswer)
          ) {
            allTrue = false;
          }
        }
        return allTrue;
      })
      .every(allTrue => allTrue);
    const question = await this.questionRepository.findOne(dto.questionId);
    const selectedAnswersArray = dto.subQuestions.map(subQuestion => ({
      questionId: subQuestion.subQuestionId,
      selectedAnswers: subQuestion.selectedAnswers,
    }));
    return {
      pass,
      question,
      selectedAnswersArray,
    };
  }

  async estimateThirdQuestion(dto: EstimateThirdQuestionDto) {
    return await this.estimateSecondQuestion(dto);
  }

  async getFileNParse(): Promise<
    Array<
      CreateCommonQuestionDto | CreateSecondQuestionDto | CreateThirdQuestionDto
    >
  > {
    const newDir = path.join(process.cwd(), './src/questionsInput/new');
    const filePath = type => name =>
      path.join(process.cwd(), `./src/questionsInput/${type}/${name}`);
    const fileList = fs.readdirSync(newDir);
    const files = fileList.map(name =>
      fs.readFileSync(filePath('new')(name), { encoding: 'utf8' }),
    );
    const time = new Date().getTime();
    const savedDir = path.join(
      process.cwd(),
      `./src/questionsInput/saved/${time}`,
    );
    fs.mkdirSync(savedDir);
    for (const name of fileList) {
      fs.copyFileSync(filePath('new')(name), filePath(`saved/${time}`)(name));
      fs.unlinkSync(filePath('new')(name));
    }
    return parseFiles(files);
  }

  async saveAllQuestions(
    dto: Array<
      CreateCommonQuestionDto | CreateSecondQuestionDto | CreateThirdQuestionDto
    >,
  ) {
    const questions = [];
    for (const question of dto) {
      switch (question.type) {
        case 1:
          const savedQuestion1 = await this.saveCommonQuestion(question);
          questions.push(savedQuestion1);
          break;
        case 2:
          const savedQuestion2 = await this.saveSecondQuestion(question);
          questions.push(savedQuestion2);
          break;
        case 3:
          const savedQuestion3 = await this.saveThirdQuestion(question);
          questions.push(savedQuestion3);
      }
    }
    return questions;
  }

  private async saveAnswer(content: string[]) {
    const answers = content.map(c => {
      const answer = new Answer();
      answer.content = c;
      return answer;
    });
    return await this.anwserRepository.save(answers);
  }
}
