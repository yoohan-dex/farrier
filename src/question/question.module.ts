import { Module } from '@nestjs/common';
import { QuestionService } from './question.service';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Question } from '../entity/Question';
import { Answer } from '../entity/Answer';
import { QuestionController } from './question.controller';
import { FailAnswer } from '../entity/FailAnswer';
import { Performance } from '../entity/Performance';
import { PerformanceService } from '../performance/performance.service';

@Module({
  imports: [
    TypeOrmModule.forFeature([Question, Answer, Performance, FailAnswer]),
  ],
  providers: [QuestionService, PerformanceService],
  controllers: [QuestionController],
})
export class QuestionModule {}
