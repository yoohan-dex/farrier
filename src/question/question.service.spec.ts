import { Test, TestingModule } from '@nestjs/testing';
import { QuestionService } from './question.service';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Question } from '../entity/Question';
import { Answer } from '../entity/Answer';
import {
  CreateSecondQuestionDto,
  CreateThirdQuestionDto,
  CreateSubQuestionDto,
  CreateCommonQuestionDto,
  EstimateCommonQuestionDto,
} from './question.dto';

const mockCommonQuestion: CreateCommonQuestionDto = {
  type: 1,
  content: 'this is a question and 中文',
  answers: ['answer1', 'answer2', 'answer3', 'answer4'],
  rightAnswer: ['answer3'],
  analysis: 'then i am analysis',
};

const subCommonQuestion: CreateSubQuestionDto = {
  type: 0,
  content: 'this is a question and 中文',
  answers: ['answer1', 'answer2', 'answer3', 'answer4'],
  rightAnswer: ['answer3'],
  analysis: 'then i am analysis',
};

const mockSecondQuestion: CreateSecondQuestionDto = {
  type: 2,
  content: 'whatever you write',
  questions: [subCommonQuestion, subCommonQuestion],
};

const mockThirdQuestion: CreateThirdQuestionDto = {
  type: 3,
  answers: ['answer1', 'answer2', 'answer3', 'answer4'],
  questions: [subCommonQuestion, subCommonQuestion],
};

describe('QuestionService', () => {
  let service: QuestionService;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        TypeOrmModule.forRoot(),
        TypeOrmModule.forFeature([Question, Answer]),
      ],
      providers: [QuestionService],
    }).compile();

    service = module.get<QuestionService>(QuestionService);
  });
  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should can save a new common question', async () => {
    const commonQuestionMatch = {
      id: expect.any(Number),
      type: mockCommonQuestion.type,
      answers: mockCommonQuestion.answers.map(answer => ({
        id: expect.any(Number),
        content: answer,
      })),
      parent: null,
      content: mockCommonQuestion.content,
      analysis: mockCommonQuestion.analysis,
      rightAnswer: [expect.any(Number)],
    };
    const question = await service.saveCommonQuestion(mockCommonQuestion);
    expect(question).toMatchSnapshot(commonQuestionMatch);
  });

  it('should can save a new second question', async () => {
    const subQuestionMatch = {
      id: expect.any(Number),
      type: subCommonQuestion.type,
      answers: subCommonQuestion.answers.map(answer => ({
        id: expect.any(Number),
        content: answer,
      })),
      content: subCommonQuestion.content,
      analysis: subCommonQuestion.analysis,
      rightAnswer: [expect.any(Number)],
    };
    const question = await service.saveSecondQuestion(mockSecondQuestion);
    expect(question).toMatchSnapshot({
      id: expect.any(Number),
      type: mockSecondQuestion.type,
      content: mockSecondQuestion.content,
      questions: [subQuestionMatch, subQuestionMatch],
      analysis: null,
      rightAnswer: null,
    });
  });

  it('should can save a new third question', async () => {
    const subQuestionMatch = {
      id: expect.any(Number),
      type: subCommonQuestion.type,
      content: subCommonQuestion.content,
      analysis: subCommonQuestion.analysis,
      rightAnswer: [expect.any(Number)],
    };
    const question = await service.saveThirdQuestion(mockThirdQuestion);
    expect(question).toMatchSnapshot({
      id: expect.any(Number),
      type: mockThirdQuestion.type,
      answers: mockThirdQuestion.answers.map(answer => ({
        id: expect.any(Number),
        content: answer,
      })),
      questions: [subQuestionMatch, subQuestionMatch],
      content: null,
      rightAnswer: null,
      analysis: null,
    });
  });

  it('should can get questions by type and by pager', async () => {
    const subQuestionMatch = {
      id: expect.any(Number),
      type: subCommonQuestion.type,
      content: subCommonQuestion.content,
      analysis: subCommonQuestion.analysis,
      rightAnswer: [expect.any(Number)],
      answers: [],
    };
    const questions = await service.getQuestionByType(3, 1, 1);
    expect(questions).toMatchSnapshot([
      {
        id: expect.any(Number),
        type: 3,
        answers: mockThirdQuestion.answers.map(answer => ({
          id: expect.any(Number),
          content: answer,
        })),
        questions: [subQuestionMatch, subQuestionMatch],
        content: null,
        rightAnswer: null,
        analysis: null,
      },
    ]);
  });

  it('will estimate a common question', async () => {
    const [question] = await service.getQuestionByType(1, 1, 1);
    const dto: EstimateCommonQuestionDto = {
      questionId: question.id,
      selectedAnswers: [8],
      type: 1,
    };
    const result = await service.estimateCommonQuestion(dto);
    expect(result).toMatchSnapshot({
      pass: false,
      question: {
        id: expect.any(Number),
        type: mockCommonQuestion.type,
        answers: mockCommonQuestion.answers.map(answer => ({
          id: expect.any(Number),
          content: answer,
        })),
        content: mockCommonQuestion.content,
        analysis: mockCommonQuestion.analysis,
        rightAnswer: [expect.any(Number)],
      },
      selectedAnswer: [
        {
          questionId: expect.any(Number),
          selectedAnswers: [8],
        },
      ],
    });
  });
});
