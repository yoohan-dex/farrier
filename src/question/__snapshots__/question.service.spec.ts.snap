// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionService should can get questions by type and by pager 1`] = `
Array [
  Object {
    "analysis": null,
    "answers": Array [
      Object {
        "content": "answer1",
        "id": Any<Number>,
      },
      Object {
        "content": "answer2",
        "id": Any<Number>,
      },
      Object {
        "content": "answer3",
        "id": Any<Number>,
      },
      Object {
        "content": "answer4",
        "id": Any<Number>,
      },
    ],
    "content": null,
    "id": Any<Number>,
    "questions": Array [
      Object {
        "analysis": "then i am analysis",
        "answers": Array [],
        "content": "this is a question and 中文",
        "id": Any<Number>,
        "rightAnswer": Array [
          Any<Number>,
        ],
        "type": 0,
      },
      Object {
        "analysis": "then i am analysis",
        "answers": Array [],
        "content": "this is a question and 中文",
        "id": Any<Number>,
        "rightAnswer": Array [
          Any<Number>,
        ],
        "type": 0,
      },
    ],
    "rightAnswer": null,
    "type": 3,
  },
]
`;

exports[`QuestionService should can save a new common question 1`] = `
Question {
  "analysis": "then i am analysis",
  "answers": Array [
    Object {
      "content": "answer1",
      "id": Any<Number>,
    },
    Object {
      "content": "answer2",
      "id": Any<Number>,
    },
    Object {
      "content": "answer3",
      "id": Any<Number>,
    },
    Object {
      "content": "answer4",
      "id": Any<Number>,
    },
  ],
  "content": "this is a question and 中文",
  "id": Any<Number>,
  "parent": null,
  "rightAnswer": Array [
    Any<Number>,
  ],
  "type": 1,
}
`;

exports[`QuestionService should can save a new second question 1`] = `
Question {
  "analysis": null,
  "content": "whatever you write",
  "id": Any<Number>,
  "questions": Array [
    Object {
      "analysis": "then i am analysis",
      "answers": Array [
        Object {
          "content": "answer1",
          "id": Any<Number>,
        },
        Object {
          "content": "answer2",
          "id": Any<Number>,
        },
        Object {
          "content": "answer3",
          "id": Any<Number>,
        },
        Object {
          "content": "answer4",
          "id": Any<Number>,
        },
      ],
      "content": "this is a question and 中文",
      "id": Any<Number>,
      "rightAnswer": Array [
        Any<Number>,
      ],
      "type": 0,
    },
    Object {
      "analysis": "then i am analysis",
      "answers": Array [
        Object {
          "content": "answer1",
          "id": Any<Number>,
        },
        Object {
          "content": "answer2",
          "id": Any<Number>,
        },
        Object {
          "content": "answer3",
          "id": Any<Number>,
        },
        Object {
          "content": "answer4",
          "id": Any<Number>,
        },
      ],
      "content": "this is a question and 中文",
      "id": Any<Number>,
      "rightAnswer": Array [
        Any<Number>,
      ],
      "type": 0,
    },
  ],
  "rightAnswer": null,
  "type": 2,
}
`;

exports[`QuestionService should can save a new third question 1`] = `
Question {
  "analysis": null,
  "answers": Array [
    Object {
      "content": "answer1",
      "id": Any<Number>,
    },
    Object {
      "content": "answer2",
      "id": Any<Number>,
    },
    Object {
      "content": "answer3",
      "id": Any<Number>,
    },
    Object {
      "content": "answer4",
      "id": Any<Number>,
    },
  ],
  "content": null,
  "id": Any<Number>,
  "questions": Array [
    Object {
      "analysis": "then i am analysis",
      "content": "this is a question and 中文",
      "id": Any<Number>,
      "rightAnswer": Array [
        Any<Number>,
      ],
      "type": 0,
    },
    Object {
      "analysis": "then i am analysis",
      "content": "this is a question and 中文",
      "id": Any<Number>,
      "rightAnswer": Array [
        Any<Number>,
      ],
      "type": 0,
    },
  ],
  "rightAnswer": null,
  "type": 3,
}
`;

exports[`QuestionService will estimate a common question 1`] = `
Object {
  "pass": false,
  "question": Object {
    "analysis": "then i am analysis",
    "answers": Array [
      Object {
        "content": "answer1",
        "id": Any<Number>,
      },
      Object {
        "content": "answer2",
        "id": Any<Number>,
      },
      Object {
        "content": "answer3",
        "id": Any<Number>,
      },
      Object {
        "content": "answer4",
        "id": Any<Number>,
      },
    ],
    "content": "this is a question and 中文",
    "id": Any<Number>,
    "rightAnswer": Array [
      Any<Number>,
    ],
    "type": 1,
  },
  "selectedAnswer": Array [
    Object {
      "questionId": Any<Number>,
      "selectedAnswers": Array [
        8,
      ],
    },
  ],
}
`;
